src/libos/src/process/syscalls.rs

```diff
@@ -247,11 +255,31 @@ pub fn do_getppid() -> Result<isize> {
     Ok(ppid as isize)
 }
 
-pub fn do_getpgid() -> Result<isize> {
-    let pgid = super::do_getpid::do_getpgid();
+pub fn do_getpgid(pid: pid_t) -> Result<isize> {
+    let real_pid = if pid == 0 { do_getpid()? as pid_t } else { pid };
+    let pgid = super::pgrp::do_getpgid(real_pid)?;
     Ok(pgid as isize)
 }
 
+pub fn do_setpgid(pid: i32, pgid: pid_t) -> Result<isize> {
+    if pid < 0 {
```

Comment:
Due to previous comments:
EINVAL: pgid is less than 0 (setpgid(), setpgrp()).
ESRCH: For getpgid(): pid does not match any process. For setpgid(): pid is not the calling process and not a child of the calling process.

Reply:
Done.